src/FTPObject.jl

```diff
@@ -132,168 +132,117 @@ end
-                   file and false for any that failed to transfer.
+`FTPResponse`: Returns the ftp response object, else nothing
 """
 function upload(
     ftp::FTP,
-    local_file_paths::Vector{<:AbstractString},
-    ftp_dir::AbstractString;
-    retry_callback::Function=(count, options) -> (count < 4, options),
-    retry_wait_seconds::Integer=5,
-    verbose=nothing,
+    local_path::AbstractString,
+    remote_path::AbstractString;
+    ftp_options=ftp.ctxt,
+    mode::FTP_MODE=binary_mode,
+    verbose=nothing
 )
     _dep_verbose_kw(verbose, typeof(ftp), :upload)
-    successful_delivery = Bool[]
-
-    ftp_options = ftp.ctxt
-
-    for single_file in local_file_paths
-        # The location we are going to drop the file in the FTP server
-        server_location = joinpath(ftp_dir, basename(single_file))
-
-        # ftp_put requires an io so open up our file.
-        open(single_file) do single_file_io
-            # Whether or not the current file was successfully delivered to the FTP
-            file_delivery_success = false
-
-            attempts = 1
-            # The loops should break after an appropriate amount of retries.
-            # This way of doing retries makes testing easier.
-            # Defaults to 4 attempts, waiting 5 seconds between each retry.
-            while true
-                try
-                    resp = ftp_put(
-                        ftp_options, server_location, single_file_io; verbose=verbose
-                    )
-                    file_delivery_success = resp.code == complete_transfer_code
-                    if file_delivery_success
-                        break
-                    end
-                catch e
-                    @warn(e)
-                end
-                sleep(retry_wait_seconds)
-                # It returns ftp_options for testing purposes, where the ftp server
-                # starts not existing then comes into existence during retries.
-                do_retry, ftp_options = retry_callback(attempts, ftp_options)
-                if !do_retry
-                    break
-                end
-                attempts += 1
-            end
-
-            push!(successful_delivery, file_delivery_success)
 
-        end
+    # The location we are going to drop the file in the FTP server
+    if basename(remote_path) == ""
+        # If the remote path was just a directory, then the full remote path should be
+        # that directory plus the basename of the local_path
+        server_location = joinpath(dirname(remote_path), basename(local_path))
+    elseif remote_path == "."
+        # Handle the special case where remote_path is "."
+        # Since the dirname/basename strategy doesn't work in this specific case, we just
+        # handle it here.
+        server_location = joinpath(".", basename(local_path))
+    else
+        # If the remote path is a full file path, then just use that
+        server_location = remote_path
     end
```

Comment:
What if the remote path is `a/b/.`?

Reply:
good point. I guess the case is if the last character is '.' which I have looked at.

I suppose one problem is that there isn't really a difference between directories and filenames as far as this function is concerned. It assumes that if a remote path ends with '/' or '.' it is a directory, but aside from that it has no idea. Is it ok to leave that up to the user, or should we be doing a thing where we detect if the remote is a dir or not and act accordingly?

Reply:
That sounds sensible to me. If someone tries to write a file with the same name as a directory we should let LibCURL sort that out.

Reply:
ok fair enough.

Reply:
I think you want `basename(remote_path) == "."` otherwise `a/b/c.` would work for `remote_path[end] == '.'`

Reply:
Another missing case: `a/b/..`

I think you can cover all of these cases with:
```julia
if basename(remote_path) in ("", ".", "..")
    server_location = joinpath(remote_path, basename(local_path))
else
    server_location = remote_path
end
```